internal/partitions/partitions.go

```diff
@@ -329,13 +384,82 @@ func GetSizeAndDocumentCount(ctx context.Context, logger *logger.Logger, retryer
 	return value.Size, value.Count, value.Capped, nil
 }
 
-// getNumPartitions returns the total number of partitions needed for the collection.
+// GetDocumentCountAfterFiltering counts the number of filtered documents in a collection.
+//
+// This function could take a long time, especially if the collection does not have an index
+// on the filtered fields.
+func GetDocumentCountAfterFiltering(ctx context.Context, logger *logger.Logger, retryer *retry.Retryer, srcColl *mongo.Collection, collUUID util.UUID, filter map[string]any) (int64, error) {
+	srcDB := srcColl.Database()
+	collName := srcColl.Name()
+
+	value := struct {
+		Count int64 `bson:"numFilteredDocs"`
+	}{}
+
+	var pipeline mongo.Pipeline
+
+	if len(filter) > 0 {
+		pipeline = append(pipeline, bson.D{{"$match", filter}})
+	}
+	pipeline = append(pipeline, []bson.D{
+		{{"$count", "numFilteredDocs"}},
+	}...)
+
+	currCollName, err := retryer.RunForUUIDAndTransientErrors(ctx, logger, collName, func(ri *retry.Info, collectionName string) error {
+		ri.Log(logger.Logger, "count", "source", srcDB.Name(), collectionName, "Counting filtered documents.")
+		request := retryer.RequestWithUUID(bson.D{
+			{"aggregate", collectionName},
+			{"pipeline", pipeline},
+			{"cursor", bson.D{}},
+		}, collUUID)
+
+		cursor, driverErr := srcDB.RunCommandCursor(ctx, request)
+		if driverErr != nil {
+			return driverErr
+		}
+
+		defer cursor.Close(ctx)
+		if cursor.Next(ctx) {
+			if err := cursor.Decode(&value); err != nil {
+				return errors.Wrapf(err, "failed to decode $count response for source namespace %s.%s after filter (%+v)", srcDB.Name(), collName, filter)
+			}
+		}
+		return nil
+	})
+
+	// TODO (REP-960): remove this check.
+	// If we get NamespaceNotFoundError then return 0 since we won't do any partitioning with those returns
+	// and the aggregation did not fail so we do not want to return an error. A
+	// NamespaceNotFoundError can happen if the database does not exist.
+	if util.IsNamespaceNotFoundError(err) {
+		return 0, nil
+	}
+
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to run aggregation $count for source namespace %s.%s after filter (%+v)", srcDB.Name(), collName, filter)
+	}
+
+	// CollectionUUIDMismatch where the collection does not exist will return a nil cursor and nil
+	// error.
+	if currCollName == "" {
+		// Return 0, nil as CollectionUUIDMismatch should not cause an initial sync error.
+		return 0, nil
+	}
+
+	logger.Debug().Msgf("Collection %s.%s filtered document count: %d, filter: %+v",
+		srcDB.Name(), currCollName, value.Count, filter)
+
+	return value.Count, nil
+}
+
+// getNumPartitions returns the total number of partitions needed for the collection,
+// which is proportional to the percentage of filtered documents in the collection.
 //
 // The returned number is always 1 or greater, where 1 indicates that the collection
```

Comment:
Is the returned number aways >= 1 even with the ratio? I am guessing yes since we can't have 0 partitions?

Reply:
Yea, the return value adds 1 to a non-negative int.